Show discussion component on Team view

andy-armstrong opened the pull request:
https://openedx.atlassian.net/browse/TNL-2515

A test course is available on my sandbox:

http://andy-armstrong.m.sandbox.edx.org/courses/course-v1:AndyA+AA101+1/teams/

explorerleslie commented:
Functionality looks good, although I keep getting a "this thread is deleted" message when clicking expand/collapse on a post. Is that related to permissions model not being implemented? @andy-armstrong

andy-armstrong commented:
Thanks @explorerleslie (and awesome test video!). I'll investigate the expand/collapse issue as that shouldn't be failing.

andy-armstrong commented:
@explorerleslie It turns out you were correct that this is a permissions problem. The code currently thinks that this is an inline discussion and yet it isn't listed as one of the topics that the current user has permission to access. We'll have to decide whether to fix this in my PR or in @cahrens's permissions PR.

andy-armstrong commented:
@cahrens As we discussed, the particular line of code that is doing this permission check is:

https://github.com/edx/edx-platform/blob/master/lms/djangoapps/django_comment_client/forum/views.py#L320

andy-armstrong commented:
I've switched over to a new sandbox for this PR:

http://pr9105.m.sandbox.edx.org/courses/course-v1:AndyA+AA101+1/teams/#teams/solar/sun-lovers

andy-armstrong commented:
@cahrens @dan-f I've implemented Jasmine tests for the two new views, so please review.

cahrens commented:
@andy-armstrong Do you mind if I want to review until you fix the known issues on the sandbox? Specifically--

1) Can't respond to a post (the response field doesn't show)
2) Can't edit my own post, get "Topic doesn't exist"
3) Can change the discussion topic to something else

andy-armstrong commented:
@cahrens @peter-fogg @bderusha @explorerleslie @marcotuts @catong I've fixed the remaining reported bugs and I've updated my sandbox to have the latest code (including the correct version of the comment service):

http://pr9105.sandbox.edx.org/courses/course-v1:AndyA+AA101+1/teams/#teams/solar/solar-power

The honor user is enrolled in the Solar Power team so that you can test the non-staff behavior.

@cahrens @peter-fogg Please review.

cahrens commented on common/static/coffee/src/discussion/discussion_module_view.coffee:
```diff
@@ -10,10 +10,11 @@ if Backbone?
       @toggleDiscussionBtn = @$(".discussion-show")
       # Set the page if it was set in the URL. This is used to allow deep linking to pages
       match = @page_re.exec(window.location.href)
+      @context = options.context or "course"  # allowed values are "course" or "standalone"
```
As I mentioned on the specific commit, I'd like to talk about us doing this as an editing option instead of linking it to context (since we don't pervasively need the context in the JavaScript code).

andy-armstrong replied:
Sure thing, let's discuss it. I can imagine that we'll have lots of team-specific customizations, so having a single setting that can drive them all seems more flexible. I'm not sure I like the idea of introducing very targeted properties for individual capabilities that we want, but I'll entertain the idea.

cahrens commented:
Is the deletion of the empty file common/static/coffee/src/discussion/templates.coffee correct?

andy-armstrong commented:
@cahrens I did deliberately remove the empty file because I couldn't see any reason for its existence. Given how the files get bundled in production, I couldn't see any way in which it could make a difference.

cahrens commented on common/static/coffee/src/discussion/views/discussion_thread_edit_view.js:
```diff
@@ -75,7 +81,9 @@
                         this.$('.edit-post-body textarea').val('').attr('prev-text', '');
                         this.$('.wmd-preview p').html('');
-                        postData.courseware_title = this.topicView.getFullTopicName();
+                        if (this.topicView) {
```
Why are there 2 places where update postData gets set? Just trying to understand the existing code...

andy-armstrong replied:
It looks like postData is passed to the AJAX call, but then is subsequently updated and then set into the model. This does seem very confusing...

peter-fogg replied:
Same nit as above.

andy-armstrong commented on common/static/coffee/src/discussion/views/new_post_view.coffee:
```diff
@@ -87,7 +87,6 @@ if Backbone?
               url: url
               type: "POST"
               dataType: 'json'
-              async: false # TODO when the rest of the stuff below is made to work properly..
```
Reviewers: I removed this because sinon doesn't have a good way to handle synchronous requests, and also because I can't see how it can be a good idea for the UX. We should think through what if any implications there might be for this change.

cahrens replied:
This scares me a bit. So deleting this line means that the ajax request will be asynchronous? Can you see who wrote the TODO so we can ask for more context?

andy-armstrong replied:
It looks like Arjun Singh added this on September 1st, 2012, and he included the TODO even back then. It appears that it was a transitional thing that was never revisited. I can see why it scares you, but we don't use synchronous AJAX anywhere else that I can think of. To me, blocking the browser waiting for the response seems worse.

If we don't make this change, we need to find a way to get all the Jasmine tests to work with the synchronous call. The way ``AjaxHelpers`` works, we invoke an action and then afterwards we declare the response to be given. This flow doesn't work with synchronous calls, so we need to instead declare what the result is going to be before we invoke the action.

Interestingly, synchronous AJAX has been deprecated since JQuery 1.8, so removing it seems necessary for us to upgrade:

http://stackoverflow.com/questions/11448011/jquery-ajax-methods-async-option-deprecated-what-now

cahrens replied:
@jimabramson Do you know if there was ever an attempt to make this AJAX callback asynchronous? If so, do you have any recollection of what the issues were? In other words, can you think of any reason why we should not make this change?

cahrens commented on common/static/coffee/src/discussion/views/discussion_thread_view.coffee:
```diff
@@ -19,6 +19,7 @@ if Backbone?
     initialize: (options) ->
       super()
       @mode = options.mode or "inline"  # allowed values are "tab" or "inline"
+      @context = options.context or "course"  # allowed values are "course" or "standalone"
```
What is the relationship between discussion_module_view.coffee and this file (why does the context need to be passed through in both places)?

andy-armstrong replied:
DiscussionModuleView is the outermost view, and it makes a DiscussionThreadView per post (the terminology is very confusing). I needed to wire the context down through the children so that it can get to the DiscussionThreadEditView (the edit modal).

cahrens commented on common/test/acceptance/pages/lms/teams.py:
```diff
@@ -102,3 +103,42 @@ def header_topic_description(self):
+    """
+    def __init__(self, browser, course_id, team):
+        """
+        Set up `self.url_path` on instantiation, since it dynamically
```
I don't think this documentation is correct. It should talk about "team" instead of "topic".

andy-armstrong replied:
Good catch. It looks like it was copied directly from ``BrowseTeamPage``. I've fixed it.

cahrens commented on common/test/acceptance/pages/lms/teams.py:
```diff
@@ -102,3 +103,42 @@ def header_topic_description(self):
+        has_correct_url = self.url.endswith(self.url_path)
+        # The "teams-main" class is not unique to this view, but currently there is nothing else on the
+        # page except for the discussion module.
+        team_view_present = self.q(css='.teams-main').present
```
I'd like to see if there is something better we can look for besides .teams-main.

andy-armstrong replied:
I've introduced a ``team-profile`` class on the outermost div, and changed this test to look for it.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -455,3 +459,47 @@ def test_teams_membership(self):
+
+    def setup_thread(self):
+        """
+        Set up multiple threads on the team page by passing 'thread_count'.
```
Update documentation (it's only a single thread...).

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -455,3 +459,47 @@ def test_teams_membership(self):
+        Scenario: Team Page renders a team discussion.
+        Given I am enrolled in a course with a team configuration, a topic,
+            and a team belonging to that topic
+        When I post a thread in the team's discussion
```
Nit: this test doesn't actually post a thread as any particular user. I would say

When a thread exists in the team's discussion 
...
And I should see the existing thread

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/teams_tab_spec.js:
```diff
@@ -0,0 +1,92 @@
+define([
```
Was this file changed or just moved?

andy-armstrong replied:
It was just moved, I think. I don't know how the history got lost again... I'll try to recover it.

cahrens commented on lms/djangoapps/teams/static/teams/js/spec_helpers/team_discussion_helpers.js:
```diff
@@ -0,0 +1,126 @@
+        testTeamDiscussionID = "12345";
+
+    createMockPostResponse = function(options) {
+        return _.extend(_.extend({
```
Why the nested _.extend?

And wow, this is a lot of mock data.

andy-armstrong replied:
This is an idiom I've used before, in that we want a new dict that is then pre-populated with defaults, and then overridden with the provided options. I think I had issues when extending the default dict directly that it ended up corrupting it for subsequent calls. However, I'm not sure about that now. Also, I messed the idiom up here because it doesn't provide the initial empty dict.

Let me try it without the extra level and see if it works...

andy-armstrong replied:
Regarding the mock data: I'll try to trim it down and see what the minimum is that is required. This is the full set of data returned by a real response which seems unduly verbose.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/team_profile.js:
```diff
@@ -0,0 +1,33 @@
+ */
+;(function (define) {
+    'use strict';
+    define(['backbone', 'underscore', 'gettext', 'teams/js/views/team_discussion',
```
@symbolist have you taken a look at this code? It will need to sync up with the Team page you are creating.

symbolist replied:
@cahrens Yes, @muzaffaryousaf is working of this branch https://github.com/edx/edx-platform/pull/9183.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/team_profile.js:
```diff
@@ -0,0 +1,33 @@
+                },
+
+                render: function () {
+                    var canPostToTeam = true;  // TODO: determine this permission correctly!
```
Fix after Peter's PR merges in.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/team_profile.js:
```diff
@@ -0,0 +1,33 @@
+    'use strict';
+    define(['backbone', 'underscore', 'gettext', 'teams/js/views/team_discussion',
+            'text!teams/templates/team-profile.underscore'],
+        function (Backbone, _, gettext, TeamDiscussionView, team_template) {
```
Do we have a convention of using underscores in templates names?

andy-armstrong replied:
Good catch. I didn't mean to use underscores, so I've fixed it to be ``teamTemplate``. Just FYI, I don't name it like a class because the value is a string.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -10,259 +10,333 @@
+                        [':default', _.bind(this.routeNotFound, this)],
+                        ['topics/:topic_id', _.bind(this.browseTopic, this)],
+                        ['teams/:topic_id/:team_id', _.bind(this.browseTeam, this)],
+                        [new RegExp('^(browse)$'), _.bind(this.goToTab, this)],
```
I didn't review the earlier PRs... I'd like someone to explain to me how these work.

andy-armstrong replied:
The short answer is that these define the parts of the URL that appear after the ``#``. When the browser requests to navigate to such a URL, the Backbone router will invoke the specified callback. This allows the browser to provide bookmarkable URLs for different conceptual parts of a single page app.

More details: http://backbonejs.org/#Router

cahrens commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -10,259 +10,333 @@
+                        router.route.apply(router, route);
+                    });
+                    // TODO replace this with actual views!
+                    TempTabView = Backbone.View.extend({
```
This TempTavView is still needed?

andy-armstrong replied:
Yes, this is providing the fake "My Teams" tab. It will be replaced when that tab is implemented.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -10,259 +10,333 @@
+                    _.each([
+                        [':default', _.bind(this.routeNotFound, this)],
+                        ['topics/:topic_id', _.bind(this.browseTopic, this)],
+                        ['teams/:topic_id/:team_id', _.bind(this.browseTeam, this)],
```
Why "browseTeam"? Isn't this the page for a specific team?

andy-armstrong replied:
You are browsing from the list of teams into a single team, from which you can see the discussions, members etc. It is singular to indicate this, as with the previous line's ``browseTopic`` which lets you browse the contents of a single topic. 

We've always had trouble with naming the list view versus the individual view. Having an extra ``s`` to indicate the plural is confusing, but I'm not sure I like the ``ListView`` vs ``DetailView`` convention either. Let's come to an agreement as to how to name this and then I can do the renaming.

andy-armstrong replied:
I named the 'individual team' page the ``TeamProfileView`` to mirror the learner profile, but I wasn't 100% happy with that choice either...

cahrens commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -10,259 +10,333 @@
-                       );
-                   },
+                /**
+                 * Render the list of teams for the given topic ID.
```
Documentation is wrong.

And why is this browseTeam? Aren't we talking about a single team here?

cahrens commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -10,259 +10,333 @@
+                    var router = this.router,
+                        breadcrumbs, headerView;
+                    breadcrumbs = [{
+                        title: 'All Topics',
```
i18n

andy-armstrong replied:
Whoops, that slipped through from an earlier PR. I've fixed it now.

cahrens commented on lms/static/js/spec/main.js:
```diff
@@ -520,19 +555,169 @@
+                exports: 'DiscussionUtil',
+                init: function() {
+                    // Set global variables that the discussion code is expecting to be defined
+                    window.Backbone = require('backbone');
```
Oh, this is nice (as nice as one could hope for).

cahrens commented:
@andy-armstrong This is looking great! Hopefully tomorrow @peter-fogg's PR and yours can meet, and we can fix the last little bit.

@symbolist do you think it would be helpful to have someone from your team review this PR to share knowledge? I don't want to take away time from your own commitments, but more context about the changes we are making would likely be helpful for your team.

symbolist commented:
@cahrens Sure, I think @muzaffaryousaf will be the right person to do the review. He can be tagged whenever this is ready.

andy-armstrong commented:
jenkins run bokchoy

peter-fogg commented on common/static/coffee/src/discussion/views/discussion_thread_edit_view.js:
```diff
@@ -53,13 +58,14 @@
-                        commentable_id: commentableId
+                        body: body
                     };
+                if (this.topicView) {
```
Nit: could this conditional be `if ( this.context === 'course')`? As far as I can tell it's logically equivalent and makes it a bit more clear.

peter-fogg commented on common/test/acceptance/pages/lms/teams.py:
```diff
@@ -102,3 +103,38 @@ def header_topic_description(self):
         return self.q(css='.team-card')
+
+
+class TeamPage(CoursePage, PaginatedUIMixin):
```
I wish we'd standardized "Teams" versus "Team" at some point...

peter-fogg commented on lms/djangoapps/django_comment_client/base/views.py:
```diff
@@ -250,11 +250,13 @@ def update_thread(request, course_id, thread_id):
-        else:
+        thread_context = getattr(thread, "context", "course")
+        if thread_context == "course" and not discussion_category_id_access(course, request.user, commentable_id):
             return JsonError(_("Topic doesn't exist"))
```
I realize this is old code, but would it make sense to return a 404 instead of the default 400 here?

andy-armstrong replied:
Agreed, that would make more sense. I don't want to mess with this since we'll be moving to the new DRF-based API before too long, and who knows what the fall out might be.

peter-fogg commented on lms/djangoapps/django_comment_client/forum/tests.py:
```diff
@@ -336,8 +337,8 @@ class SingleThreadQueryCountTestCase(ModuleStoreTestCase):
-        (ModuleStoreEnum.Type.mongo, 1, 7, 5, 14, 8),
-        (ModuleStoreEnum.Type.mongo, 50, 7, 5, 14, 8),
+        (ModuleStoreEnum.Type.mongo, 1, 6, 4, 14, 8),
+        (ModuleStoreEnum.Type.mongo, 50, 6, 4, 14, 8),
```
Out of curiosity, what made these query counts go down?

cahrens replied:
I changed it to use the cached discussion topic method.

peter-fogg commented on lms/djangoapps/teams/static/teams/js/spec/teams_factory_spec.js:
```diff
@@ -2,17 +2,17 @@ define(["jquery", "backbone", "teams/js/teams_tab_factory"],
-                    topics_url: '',
-                    teams_url: '',
+                    topicsUrl: '',
+                    teamsUrl: '',
```
:trophy:

peter-fogg commented on lms/djangoapps/teams/static/teams/js/spec/teams_tab_spec.js:
```diff
@@ -1,80 +0,0 @@
-define([
```
Any idea why this is marked as deleted -> added instead of moved like the other files?

peter-fogg replied:
Oh, I see that @cahrens also commented on the added version.

explorerleslie commented:
@andy-armstrong things look fine to me. Did we want to stick a placeholder right-hand sidebar in there in this PR so we know sort of what it looks like when we merge the read-only teams page PR that Lahore has, or did we decide to skip that?

andy-armstrong commented:
@explorerleslie Thanks for the review. We decided to keep this simple and add the two column layout in the Lahore PR.

muhammad-ammar commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -1,192 +1,191 @@
+                    }
+
+                    this.teamModel.save(data, { wait: true })
+                        .done(function(result) {
```
@andy-armstrong awesome for using promises.

muhammad-ammar commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -7,304 +7,380 @@
+                        [':default', _.bind(this.routeNotFound, this)],
+                        ['topics/:topic_id(/)', _.bind(this.browseTopic, this)],
+                        ['topics/:topic_id/create-team(/)', _.bind(this.newTeam, this)],
+                        ['teams/:topic_id/:team_id', _.bind(this.browseTeam, this)],
```
@andy-armstrong do we need to handle slash here like we are doing for other endpoints?

andy-armstrong replied:
Hmm, that must be new since I started this work. I've added the slash handling to match.

muhammad-ammar commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -7,304 +7,380 @@
+                            }),
+                            main: new TeamEditView({
+                                tagName: 'create-new-team',
+                                collection: this.teamsCollection,
```
@andy-armstrong edit team view(edit_team.js) is not using collection. Also do we suppose to use `self` instead of `this`

andy-armstrong replied:
Good catch. I started down that path (so that the new team could be added to the collection) but then I didn't need it (and it can't have worked given the ``this``!). I've removed it now.

cahrens commented:
:+1:

muhammad-ammar commented:
@andy-armstrong Other than some inline comments changes looks good to me.

muhammad-ammar commented:
@andy-armstrong :+1:

explorerleslie commented:
:+1:

andy-armstrong commented:
jenkins run lettuce
